verification/tpm.go

```diff
@@ -0,0 +1,241 @@
+// Licensed under the Apache-2.0 license
+
+package verification
+
+import (
+	"crypto/ecdsa"
+	"crypto/sha256"
+	"crypto/sha512"
+	"crypto/x509"
+	"encoding/binary"
+	"flag"
+	"io"
+	"math"
+	"math/big"
+	"testing"
+
+	"github.com/google/go-tpm-tools/client"
+	"github.com/google/go-tpm-tools/simulator"
+	"github.com/google/go-tpm/legacy/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+var (
+	tpmPath = flag.String("tpm-path", "/dev/tpm0", "Path to the TPM device (character device or a Unix socket).")
+
+	handleNames = map[string][]tpm2.HandleType{
+		"all":       {tpm2.HandleTypeLoadedSession, tpm2.HandleTypeSavedSession, tpm2.HandleTypeTransient},
+		"loaded":    {tpm2.HandleTypeLoadedSession},
+		"saved":     {tpm2.HandleTypeSavedSession},
+		"transient": {tpm2.HandleTypeTransient},
+	}
+)
+
+func flushAllContexts(t *testing.T, tpm io.ReadWriteCloser) {
+	totalHandles := 0
+	for _, handleType := range handleNames["all"] {
+		handles, err := client.Handles(tpm, handleType)
+		if err != nil {
+			t.Fatalf("[FATAL]: Error getting handles %s: %v", *tpmPath, err)
+		}
+		for _, handle := range handles {
+			if err = tpm2.FlushContext(tpm, handle); err != nil {
+				t.Fatalf("[FATAL]: Error flushing handle 0x%x: %v", handle, err)
+			}
+			totalHandles++
+		}
+	}
+}
+
+func startTpmSession(t *testing.T, tpm io.ReadWriteCloser, alg tpm2.Algorithm) (tpmutil.Handle, []byte, error) {
+
+	flushAllContexts(t, tpm)
+	sessHandle, nonce, err := tpm2.StartAuthSession(tpm,
+		tpm2.HandleNull,
+		tpm2.HandleNull,
+		make([]byte, 16),
+		nil,
+		tpm2.SessionPolicy,
+		tpm2.AlgNull,
+		alg)
+
+	if err != nil {
+		t.Fatalf("[FATAL]: StartAuthSession() failed: %v", err)
+	}
+
+	return sessHandle, nonce, nil
+}
+
+func TestTpmPolicySigning(d TestDPEInstance, c DPEClient, t *testing.T) {
+	simulation := false
+	ctx := getInitialContextHandle(d, c, t, simulation)
+	var ec tpm2.EllipticCurve
+	var alg tpm2.Algorithm
+
+	profile, err := GetTransportProfile(d)
+	if err != nil {
+		t.Fatalf("Could not get profile: %v", err)
+	}
+	digestLen := profile.GetDigestSize()
+
+	if digestLen == len(SHA256Digest{0}) {
+		alg = tpm2.AlgSHA256
+		ec = tpm2.CurveNISTP256
+	} else if digestLen == len(SHA384Digest{0}) {
+		alg = tpm2.AlgSHA384
+		ec = tpm2.CurveNISTP384
+	}
+
+	//Create tpm auth session to get nonce and form label which is digest
+	tpm, err := simulator.Get()
```

Comment:
This is awesome, I didn't realize go-tpm-tools will just embed the simulator for you.

Can you add a comment explaining that is what is happening?

Reply:
@jhand, This is using TPM simulator provided by go-tpm-tools.
https://github.com/google/go-tpm-tools/tree/main/simulator
It executes TPM commands and internally uses go-tpm/legacy/tpm2.
Simulator has its own io.ReadWriter (buffer) and hence doesn't require the presence of TPM device file.